pkg/reconciler/revision/reconcile_resources.go

```diff
@@ -81,10 +81,9 @@ func (c *Reconciler) reconcileDeployment(ctx context.Context, rev *v1.Revision)
 			logger.Errorw("Error getting pods", zap.Error(err))
 			return nil
 		}
-		if len(pods.Items) > 0 {
-			// Arbitrarily grab the very first pod, as they all should be crashing
-			pod := pods.Items[0]
 
+	podsLoop:
+		for _, pod := range pods.Items {
```

Comment:
Are we sure that we will not be updated enough times to go through a reconciliation phase where actually all pods are in the same status? 
And that would signal a more permanent issue?
Afaik a revision reconciliation will be triggered due to the updates of any of the following: "Certificate, PA, K8s Deployment". :thinking:

Reply:
I am also worried about relying on the fact that on one hand we use an informer based lister for the deployments and direct api server call for pods using the go client. :thinking:

Reply:
In this scenario, I would probably be fine to not try to be as accurate as possible because it is just about the revision status.

What I tried is to properly handle the eventual consistency that Kubernetes has due to its persistence and asynchronous nature of informers and controllers. Reducing those considerations to the minimum to prevent crashes (the old code knew from the deployment spec that there should be more than zero replicas but still ensured that there is an item in the pod item list before accessing the first one) would also be okay for me here.

Reply:
> I am also worried (maybe not an issue?) about relying on the fact that on one hand we use an informer based lister for the deployments and direct api server call for pods using the go client, https://github.com/kubernetes/client-go/issues/1383#issuecomment-2439183325. 🤔

Would be independent from my changes, but still interesting as I had not considered that an issue. I am not too firm into how Knative sets up the Kubernetes client. In case of controller-runtime, any object access will immediatly be cached unless you configured controller-runtime to not cache that object type.

If that pod list call really goes to Kubernetes API and if we go back to just looking at the first item, we could at least run the list call with limit=1.

Reply:
The call is: 
```	pods, err := c.kubeclient.CoreV1().Pods(ns).List(ctx, metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(deployment.Spec.Selector)})
```
This is not informer based afaik.

Reply:
In general, Knative projects do `NOT` use controller-runtime lib to create client or use it access the Kubernetes object.
Knative uses:
```
1. Use `kubeclient` to access the kube core or app objects, if we know the info of groupkindversion.
2. Use `dynamicClient` to access the kube objects, if we do not know the info of groupkindversion, but need to figure out dynamically in the code. For example, https://github.com/knative/serving/blob/main/pkg/reconciler/autoscaling/kpa/scaler.go#L318
3. Use the generated client, to access the CR, customized/created for knative. For example, to get the route, we use https://github.com/knative/serving/blob/main/pkg/reconciler/service/service.go#L248
```